server/src/main/java/org/elasticsearch/common/util/concurrent/ThreadContext.java

```diff
@@ -140,19 +124,13 @@ public StoredContext stashContext() {
                 DEFAULT_CONTEXT.putHeaders(Map.of(Task.X_OPAQUE_ID, context.requestHeaders.get(Task.X_OPAQUE_ID)));
             threadLocal.set(threadContextStruct);
         } else {
-            threadLocal.set(null);
+            threadLocal.set(DEFAULT_CONTEXT);
```

Comment:
Your change sets the default context here, while master sets it to null and then checks for null at read time to return the default context instead. These approaches look equivalent to me and I think your change makes things simpler but I wonder whether I'm missing something?

Reply:
Prior to this change, it does make a difference in that the `CloseableThreadLocal` class does more work including executing a section of code that is synchronized across all threads when returning the initial value for the thread local unless the initial value is null; if the initial value is null, nothing gets stored and the synchronized code is skipped. However, once a value has been set the move to setting it null no longer bypasses the synchronized code. The regular ThreadLocal class always executes code when first accessing with an initial value but the code that is executed is not synchronized across all threads so we can make things simpler and just reference the default context instead of a null value.

Reply:
Thanks for explaining.